### DISCUSSION TITLE
Problems with data visualization / string conversion

### DISCUSSION THREAD

**Discussion by heinehenning** [apache/streampipes#1543] [Q&A] [answered]
Hi there,

I am trying to setup some connections with simple data visualization, but I am running into problems, when it comes to the visualization part.

I am using StreamPipes 0.91.0, started with docker-compose from installer/compose. The system is running inside WSL2 on my Windows 10 Notebook.

I created a MQTT connector and selected the option "Persist events" before starting the adapter. Messages are arriving, as can be seen in the details view on the specific adapter.

This is our original payload:
![image](https://user-images.githubusercontent.com/132096802/235647908-2e93eca2-85e0-4cad-abd6-4cf2fb611eae.png)

And this is the result after selecting measurements and timestamp:
![image](https://user-images.githubusercontent.com/132096802/235648053-8207d033-a084-44ab-b0eb-f121aa3d4724.png)

When I try to build a visualization with the data explorer, the widget is not showing any data just the message _"Found no data in the selected time range"_

It would be great to get some hints where to start with the trouble shooting.
One thing I noticed is, when checking the help on the DataSource, all fields are marked as "Text"-Fields.
![image](https://user-images.githubusercontent.com/132096802/235649595-64c12f27-4098-4a9e-a3a5-71d6b5f5a90a.png)

Could that be part of the problem? 

If found a "Field Converter" for string conversion in the StreamPipes docs but this pipeline element is not available under my pipeline elements.

Thanks in advance for your help!

**Comment by heinehenning** [comment]
I still struggle to make this ↑ work. 
My assumption is, that the pipeline can not really access the JSON object that is coming from the connector.

When I try to access the connector output with a JS-Eval-Processor, I can not manage to drill down to the actual value:

Data coming from the connector:
![image](https://user-images.githubusercontent.com/132096802/235856533-041025b7-b169-4769-8538-1f60614dfe0b.png)

Configuration of JS-Eval-Processor:
![image](https://user-images.githubusercontent.com/132096802/235856842-54b23115-2ccc-4083-89ed-91e2aba5fe3a.png)

Preview of Pipeliine:
![image](https://user-images.githubusercontent.com/132096802/235856951-01d18ab3-1ac2-494e-ae22-8cb6ff4f3c46.png)

Any Ideas how to access the actual values?

**Comment by tenthe** [comment]
Hello @heinehenning,
thank you very much for reporting this behavior. I will look into it and get back to you.

Cheers,
Philipp

**Reply by heinehenning** [reply]
Hi @tenthe,

thanks for checking in!


One more thing, that I have noticed:

I tried to rule out the issues with our data format and tried to pipe some random values from the JS-Eval-Processor to the data lake:

![image](https://user-images.githubusercontent.com/132096802/235862082-e1e91598-1522-4668-a0d3-f49473bdf302.png)

I am getting the followint error on the data lake. This error occured already with my previous tests. 

![image](https://user-images.githubusercontent.com/132096802/235862685-85e843f9-3cfd-4c3f-8c9d-8cc17d7e2146.png)

**Reply by tenthe** [reply, marked as answer]
Hi @heinehenning,
I have tried to reproduce your problems.
I also had several problems with the nested json structure. @dominikriemer is currently working on it.
Once the changes are complete, I will try it again.

About your JavaScript function. I suspect the problem is with the data source.
I tried the same example with a different adapter and it worked in my instance. (I used the latest release 0.91.0 for the tests)
My guess is that the adapter's event schema is broken because of the nested structure issues.
Right now you can flatten the events in the schema editor by dragging the value property to the top level.

Cheers,
Philipp
